pkg/operator/controller/dns_status.go

```diff
@@ -80,66 +90,135 @@ func computeDNSStatusConditions(dns *operatorv1.DNS, clusterIP string, haveDNSDa
 			messages = append(messages, "No DNS pods are available.")
 		case intstrErr != nil:
 			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
 			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
 		case int(numberUnavailable) > maxUnavailable:
-			status = operatorv1.ConditionTrue
 			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
 			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
+
 		}
 	}
 
-	if len(degradedReasons) != 0 {
-		// if the last status was set to false within the last transitionUnchangedToleration, skip the new update
-		// to prevent frequent status flaps, and try to keep the long-lasting state (i.e. Degraded=False). See https://bugzilla.redhat.com/show_bug.cgi?id=2037190.
-		if oldCondition != nil && oldCondition.Status == operatorv1.ConditionFalse && lastTransitionTimeIsRecent(currentTime, oldCondition.LastTransitionTime.Time, transitionUnchangedToleration) {
-			return *oldCondition
+	progressing := newProgressingCondition != nil && newProgressingCondition.Status == operatorv1.ConditionTrue
+	var currentConditions []operatorv1.OperatorCondition
+	if progressing {
+		currentConditions = append(currentConditions, *newProgressingCondition)
+	}
+
+	// If the operator is Progressing=true then we don't mark it Degraded=true until we check the rest of the conditions.
+	if !progressing {
+		if len(degradedReasons) != 0 {
+			// if the last status was set to false within the last transitionUnchangedToleration, skip the new update
+			// to prevent frequent status flaps, and try to keep the long-lasting state (i.e. Degraded=False). See https://bugzilla.redhat.com/show_bug.cgi?id=2037190.
+			if oldDegradedCondition != nil && oldDegradedCondition.Status == operatorv1.ConditionFalse && lastTransitionTimeIsRecent(currentTime, oldDegradedCondition.LastTransitionTime.Time, transitionUnchangedToleration) {
+				return *oldDegradedCondition, nil
+			}
+			degradedCondition.Status = operatorv1.ConditionTrue
+			degradedCondition.Reason = strings.Join(degradedReasons, " ")
+			degradedCondition.Message = strings.Join(messages, "\n")
+		} else {
+			degradedCondition.Status = operatorv1.ConditionFalse
+			degradedCondition.Reason = "AsExpected"
+			degradedCondition.Message = "Enough DNS pods are available, and the DNS service has a cluster IP address."
 		}
-		degradedCondition.Status = status
-		degradedCondition.Reason = strings.Join(degradedReasons, "")
-		degradedCondition.Message = strings.Join(messages, "\n")
+		currentConditions = append(currentConditions, degradedCondition)
+	}
+
+	// Now that we have information on the current DNS operator status, check the conditions against what we expect
+	graceConditions, degradedConditions, requeueAfter := cond.CheckConditions(expected, currentConditions)
```

Comment:
I don't know if `CheckConditions` makes sense in cluster-dns-operator. With CIO, we have itemized "Available-type" and "Ready-type" status, e.g ([reference](https://github.com/openshift/cluster-ingress-operator/blob/55780444031714fc931d90af298a4b193888977a/pkg/operator/controller/ingress/status.go#L547-L583)):
- IngressControllerDeploymentAvailableConditionType
- IngressControllerDeploymentReplicasMinAvailableConditionType
- IngressControllerDeploymentReplicasAllAvailableConditionType
- LoadBalancerReadyIngressConditionType
- DNSReadyIngressConditionType

Here in dns operator, we don't have these itemized conditions. In `computeDNSDegradedCondition`, we compute the reasons for each condition *not* using the status of other conditions (the old-fashion way if you will).

In the future, `CheckConditions` may make sense we if itemize the conditions, but, right now, the only situation that can possibly make sense is using `CheckConditions` with the `Available` condition. In other words, set `Degraded` based on the state of `Available`. However, we already effectively do that. The `Available` condition is based on:
1. Daemonset is created
2. There's more than 1 pod available
3. len(clusterIP) is not 0

But our `Degraded` condition already does exactly the same calculations as `Available`, but more, in `computeDNSDegradedCondition`. All this to say: DNS Conditions have been messy and I'm struggling to see the value add of using `CheckCondition` right at the moment. Feel free let me know if I'm missing the value added here.

Reply:
One reason is that it allows integration with grace conditions and the `requeueAfter` duration.  Right now `CheckConditions` is only used in `computeDNSDegraded` but now have the framework in place if we want to enhance/expand in the future.

Reply:
I can understand the motivation for expansion, that's valid. However, regarding my other comments, `GracePeriod` is ineffective/dead for `Progressing` and `Degraded`, which seems a bit confusing to me.

For other new conditions in the future that we want to base degraded on, yes, they could utilize `GracePeriod`, but at the moment, the grace period logic is being added and not being unit tested because it has no use case.

My concern, in my humble opinion, is I feel like it might be a bit too early to try to standardize this logic without real DNS status condition refactoring or use cases that we can test our new logic.